<a href="https://colab.research.google.com/github/daffamilliano/Data-Mining/blob/main/Apriori.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Menganalisa Pembelian Barang Menggunakan Algoritma Apriori**



Teman saya yang bernama Joko ingin membuat acara diskon besar-besaran pada toko online nya. Namun, Ia bingung dengan barang apa yang sering dibeli para pembelinya. Disini saya memberitahu kepada Joko. "Sepertinya masalah ini bisa dipecahkan dengan algoritma apriori deh" Ucap saya. Lalu Joko meminta saya untuk memecahkan masalahnya tersebut.

In [ ]:
# Pertama kita load dataset transaksi barang tersebut
df = pd.read_csv('drive/MyDrive/Dataset/Apriori.csv')
df.head()

ID_Transaksi                                   Barang
0        A0001  Broccoli, Green Peppers, Corn, Broccoli
1        A0002        Asparagus, Squash, Corn, Tomatoes
2        A0003            Corn, Tomatoes, Beans, Squash
3        A0004    Green Peppers, Corns, Tomatoes, Beans
4        A0005  Broccoli, Corn, Tomatoes, Green Peppers

In [ ]:
# Menginstall alat tempur untuk memecahkan masalah joko
!pip install apyori
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from apyori import apriori

Ketika kita sudah menginstall alat tempur, kita harus menghancurkan lawannya alias memecahkan masalahnya.

In [ ]:
# Membuat "data" untuk tabel yang hanya berisi Barang dari dataset sebelumnya
data=df.drop(['ID_Transaksi'],axis=1)

In [ ]:
# Lalu kita pastikan lagi hanya tinggal barang, sebelum nya dengan kolom 2 dan baris 11
data.shape

(11, 1)

In [ ]:
# Membuat list dalam list dari transaksi pembelian barang
records = []
for i in range(data.shape[0]):
    records.append([str(data.values[i,j]).split(',') for j in range(data.shape[1])])

trx = [[] for trx in range(len(records))]
for i in range(len(records)):
    for j in records[i][0]:
        trx[i].append(j)
trx

[['Broccoli', ' Green Peppers', ' Corn', ' Broccoli'],
 ['Asparagus', ' Squash', ' Corn', ' Tomatoes'],
 ['Corn', ' Tomatoes', ' Beans', ' Squash'],
 ['Green Peppers', ' Corns', ' Tomatoes', ' Beans'],
 ['Broccoli', ' Corn', ' Tomatoes', ' Green Peppers'],
 ['Squash', ' Asparagus', ' Beans', ' Tomatoes'],
 ['Tomatoes', ' Corn', ' Green Peppers'],
 ['Broccoli', ' Tomatoes', ' Green Peppers', ' Beans'],
 ['Squash', ' Asparagus', ' Beans', ' Broccoli', ' Green Peppers'],
 ['Beans', ' Corn', ' Squash', ' Asparagus', ' Beans'],
 ['Squash', ' Tomatoes', ' Broccoli', ' Green Peppers']]

In [ ]:
association_rules = apriori(trx)
association_results = association_rules

In [ ]:
# Menampilkan hasil asosiasi dari item 
pd.set_option('max_colwidth', 1000)
Result=pd.DataFrame(columns=['Rule','Support','Confidence'])
for item in association_results:
    pair = item[2]
    for i in pair:
        items = str([x for x in i[0]])
        if i[3]!=1:
            Result=Result.append({
                'Rule':str([x for x in i[0]])+ " -> " +str([x for x in i[1]]),
                'Support':str(round(item[1]*100,2))+'%',
                'Confidence':str(round(i[2] *100,2))+'%'
                },ignore_index=True)
Result

Rule Support Confidence
0                      [' Asparagus'] -> [' Beans']  27.27%     100.0%
1                      [' Beans'] -> [' Asparagus']  27.27%      50.0%
2                      [' Asparagus'] -> ['Squash']  18.18%     66.67%
3                      ['Squash'] -> [' Asparagus']  18.18%     66.67%
4                  [' Beans'] -> [' Green Peppers']  18.18%     33.33%
5                  [' Green Peppers'] -> [' Beans']  18.18%     33.33%
6                         [' Beans'] -> [' Squash']  18.18%     33.33%
7                         [' Squash'] -> [' Beans']  18.18%     66.67%
8                       [' Beans'] -> [' Tomatoes']  36.36%     66.67%
9                       [' Tomatoes'] -> [' Beans']  36.36%     57.14%
10                         [' Beans'] -> ['Squash']  18.18%     33.33%
11                         ['Squash'] -> [' Beans']  18.18%     66.67%
12              [' Broccoli'] -> [' Green Peppers']  27.27%     100.0%
13              [' Green Peppers'] -> [' Broccoli']  27.27%      50.0%
14                      [' Broccoli'] -> ['Squash']  18.18%     66.67%
15                      ['Squash'] -> [' Broccoli']  18.18%     66.67%
16                  [' Corn'] -> [' Green Peppers']  27.27%      60.0%
17                  [' Green Peppers'] -> [' Corn']  27.27%      50.0%
18                         [' Corn'] -> [' Squash']  18.18%      40.0%
19                         [' Squash'] -> [' Corn']  18.18%     66.67%
20                       [' Corn'] -> [' Tomatoes']  18.18%      40.0%
21                       [' Tomatoes'] -> [' Corn']  18.18%     28.57%
22                        [' Corn'] -> ['Broccoli']  18.18%      40.0%
23                        ['Broccoli'] -> [' Corn']  18.18%     66.67%
24              [' Green Peppers'] -> [' Tomatoes']  27.27%      50.0%
25              [' Tomatoes'] -> [' Green Peppers']  27.27%     42.86%
26               [' Green Peppers'] -> ['Broccoli']  27.27%      50.0%
27               ['Broccoli'] -> [' Green Peppers']  27.27%     100.0%
28                 [' Green Peppers'] -> ['Squash']  18.18%     33.33%
29                 ['Squash'] -> [' Green Peppers']  18.18%     66.67%
30                     [' Squash'] -> [' Tomatoes']  18.18%     66.67%
31                     [' Tomatoes'] -> [' Squash']  18.18%     28.57%
32                    [' Tomatoes'] -> ['Broccoli']  18.18%     28.57%
33                    ['Broccoli'] -> [' Tomatoes']  18.18%     66.67%
34                      [' Tomatoes'] -> ['Squash']  18.18%     28.57%
35                      ['Squash'] -> [' Tomatoes']  18.18%     66.67%
36           [' Asparagus'] -> [' Beans', 'Squash']  18.18%     66.67%
37           [' Beans'] -> [' Asparagus', 'Squash']  18.18%     33.33%
38           ['Squash'] -> [' Asparagus', ' Beans']  18.18%     66.67%
39           [' Asparagus', ' Beans'] -> ['Squash']  18.18%     66.67%
40           [' Asparagus', 'Squash'] -> [' Beans']  18.18%     100.0%
41           ['Squash', ' Beans'] -> [' Asparagus']  18.18%     100.0%
42    [' Broccoli'] -> [' Green Peppers', 'Squash']  18.18%     66.67%
43    [' Green Peppers'] -> ['Squash', ' Broccoli']  18.18%     33.33%
44    ['Squash'] -> [' Green Peppers', ' Broccoli']  18.18%     66.67%
45    [' Green Peppers', ' Broccoli'] -> ['Squash']  18.18%     66.67%
46    ['Squash', ' Broccoli'] -> [' Green Peppers']  18.18%     100.0%
47    [' Green Peppers', 'Squash'] -> [' Broccoli']  18.18%     100.0%
48      [' Corn'] -> [' Green Peppers', 'Broccoli']  18.18%      40.0%
49      [' Green Peppers'] -> [' Corn', 'Broccoli']  18.18%     33.33%
50      ['Broccoli'] -> [' Green Peppers', ' Corn']  18.18%     66.67%
51      [' Green Peppers', ' Corn'] -> ['Broccoli']  18.18%     66.67%
52      [' Corn', 'Broccoli'] -> [' Green Peppers']  18.18%     100.0%
53      [' Green Peppers', 'Broccoli'] -> [' Corn']  18.18%     66.67%
54  [' Green Peppers'] -> ['Broccoli', ' Tomatoes']  18.18%     33.33%
55  [' Tomatoes'] -> [' Green Peppers', 'Broccoli']  18.18%     28.57%


dari daftar diatas kita dapat mencari nilai confidence nya paling besar yaitu :

[' Asparagus'] -> [' Beans']	27.27%	100.0%

[' Broccoli'] -> [' Green Peppers']	27.27%	100.0%

[' Asparagus', 'Squash'] -> [' Beans']	18.18%	100.0%

['Squash', ' Broccoli'] -> [' Green Peppers']	18.18%	100.0%

[' Green Peppers', 'Squash'] -> [' Broccoli']	18.18%	100.0%

[' Corn', 'Broccoli'] -> [' Green Peppers']	18.18%	100.0%

['Broccoli', ' Tomatoes'] -> [' Green Peppers']	18.18%	100.0%

Kita dapat melihat bahwa ketika orang membeli Asparagus berarti mereka juga membeli Beans dan seterusnya.

Masalah joko pun terselesaikan dan tidak lagi bingung untuk menjual barang yang ingin ia diskon.